Solve the following problem using [Python SciPy.optimize][]. Please attach your code and
results. Specify your initial guesses of the solution. If you change
your initial guess, do you find different solutions? (**100 points**)

$$
\begin{aligned}
&\text{minimize:} && (x_1-x_2)^2 + (x_2+x_3-2)^2 + (x_4-1)^2+(x_5-1)^2 \\\\
&\text{subject to:} && x_1 + 3x_2 = 0 \\\\
&&& x_3 + x_4 - 2x_5 = 0 \\\\
&&& x_2 - x_5 = 0 \\\\
&&& -10 \leq x_i \leq 10, ~i=1,\ldots,5
\end{aligned}$$

**Note**:

1.  Please learn how to use **break points** to debug. **I will not
    address your programming questions if you have not learned how to
    debug your code.**

2.  I recommend [PyCharm][] as the IDE. If you are new to Python, you can also start with [Google Colab][] without installing anything.
    
3.  If you are on Windows, the [Anaconda][] version of Python 3 is highly recommended.


**Here are the steps to push a homework submission**:

1.  Clone the [course repo][]: First click on **Code** to get the
 Git address (e.g., the HTTPS address). Then use your IDE to clone (download) the repo using this address. 
 [PyCharm tutorial][] on using Git.

2.  You will find the homework in the **Homework** folder.

3.  For analytical problems (e.g., proofs and calculations), please use [Markdown][] to type up your answers. 
[Markdown Math][]. For Latex users, you can convert tex to markdown using [Pandoc][]. 

4. For coding problems, please submit a [Jupyter Notebook][] file with your code and final results. 
You can also add a URL to your Jupyter or Colab Notebook in README.md if you use online notebooks.

5. For each homework, please submit a single notebook file (or link) that combines the markdown solutions, 
the codes, and the computation results, and name the file according to the homework.  

6. **IMPORTANT** Please push (upload) the notebook file every time you work on the 
homework and add comments when you push, e.g., "finished problem 1, still debugging problem 2". This way I 
know you worked on your own.
 

[Python SciPy.optimize]: https://docs.scipy.org/doc/scipy/reference/tutorial/optimize.html#
[PyCharm]: https://www.jetbrains.com/pycharm/promo/?utm_source=bing&utm_medium=cpc&utm_campaign=AMER_en_US-PST%2BMST_PyCharm_Branded&utm_term=pycharm&utm_content=pycharm
[Google Colab]: https://colab.research.google.com
[Anaconda]: https://anaconda.org/anaconda/python
[course repo]: https://github.com/DesignInformaticsLab/DesignOptimization2021Fall
[PyCharm tutorial]: https://www.jetbrains.com/help/pycharm/set-up-a-git-repository.html#clone-repo
[Pandoc]: https://pandoc.org/try/
[Jupyter Notebook]: https://jupyter.org/try
[Markdown]: https://guides.github.com/features/mastering-markdown/
[Markdown Math]: http://luvxuan.top/posts/Markdown-math/

In [1]:
from scipy.optimize import minimize
# function definition
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2] - 2)**2 + (x[3] - 1)**2 + (x[4] -1)**2
# constraints
cons = ({'type': 'eq', 'fun': lambda x:  x[0] + 3*x[1]},
        {'type': 'eq', 'fun': lambda x: x[2] + x[3] - 2*x[4]},
        {'type': 'eq', 'fun': lambda x: x[1] - x[4]})
# boundaries
bnds = ((-10, 10), (-10, 10), (-10, 10), (-10, 10), (-10, 10))

In [2]:
# Result 1
# Initial guess (5,5,5,5,5)
res1 = minimize(fun, (5,5,5,5,5), method='SLSQP', bounds=bnds,
               constraints=cons) 
res1

     fun: 4.093023261962584
     jac: array([-2.04641944, -0.1860919 , -2.2325114 , -2.23267382, -1.48839509])
 message: 'Optimization terminated successfully.'
    nfev: 50
     nit: 7
    njev: 7
  status: 0
 success: True
       x: array([-0.76740731,  0.25580244,  0.62794182, -0.11633695,  0.25580244])

In [3]:
# Result 2
# Initial guess (0,0,0,0,0)
res2 = minimize(fun, (0,0,0,0,0), method='SLSQP', bounds=bnds,
               constraints=cons) 
res2

     fun: 4.093023255813954
     jac: array([-2.04651159, -0.18604648, -2.23255813, -2.23255813, -1.48837209])
 message: 'Optimization terminated successfully.'
    nfev: 23
     nit: 3
    njev: 3
  status: 0
 success: True
       x: array([-0.76744186,  0.25581395,  0.62790698, -0.11627907,  0.25581395])

In [4]:
# Result 3
# Initial guess (10,10,10,10,10)
res3 = minimize(fun, (10,10,10,10,10), method='SLSQP', bounds=bnds,
               constraints=cons) 
res3

     fun: 4.093023258012866
     jac: array([-2.04654068, -0.18606836, -2.23260909, -2.23248529, -1.48836476])
 message: 'Optimization terminated successfully.'
    nfev: 56
     nit: 8
    njev: 8
  status: 0
 success: True
       x: array([-0.76745277,  0.25581759,  0.62787786, -0.11624267,  0.25581759])



**Final solution**: x1~x5 = [-0.76745277,  0.25581759,  0.62787786, -0.11624267,  0.25581759], minimized value = 4.093023258012866

**Answer**:If changing the inital guess, the final solution is still the same. However, the iteration number is different.